examples/exti.rs

```diff
@@ -0,0 +1,68 @@
+//! Turns the user LED on
+//!
+//! Listens for interrupts on the pa7 pin. On any rising or falling edge, toggles
+//! the pc13 pin (which is connected to the LED on the blue pill board, hence the `led` name).
+
+#![no_main]
+#![no_std]
+
+use panic_halt as _;
+
+use stm32f1xx_hal::{
+    prelude::*,
+    pac
+};
+use cortex_m_rt::entry;
+use pac::interrupt;
+use core::mem::MaybeUninit;
+use embedded_hal::digital::v2::OutputPin;
+use stm32f1xx_hal::gpio::*;
+
+// These two are owned by the ISR. main() may only access them during the initialization phase,
+// where the interrupt is not yet enabled (i.e. no concurrent accesses can occur).
+// After enabling the interrupt, main() may not have any references to these objects any more.
+// For the sake of minimalism, we do not use RTFM here, which would be the better way.
+static mut LED  : MaybeUninit<stm32f1xx_hal::gpio::gpioc::PC13<Output<PushPull>>> = MaybeUninit::uninit();
+static mut EXTI : MaybeUninit<stm32f1xx_hal::gpio::gpioa::PA7<Input<Floating>>> = MaybeUninit::uninit();
+
+#[interrupt]
+fn EXTI9_5() {
+    let led = unsafe { &mut *LED.as_mut_ptr()};
+    let exti = unsafe { &mut *EXTI.as_mut_ptr()};
+
+    if exti.check_interrupt() {
+        led.toggle();
+
+        // if we don't clear this bit, the ISR would trigger indefinitely
+        exti.clear_interrupt_pending_bit();
+    }
+}
+
+#[entry]
+fn main() -> ! {
+    // initialization phase
+    let p = pac::Peripherals::take().unwrap();
+    let cp = cortex_m::peripheral::Peripherals::take().unwrap();
+    {
+        // the scope ensures that the exti reference is dropped before the first ISR can be executed.
```

Comment:
I'm assuming you do this to avoid a race condition when between `main` and the interrupt handler.

However, I don't think *holding* a reference to something can trigger a race condition, that can only happen if you actually use the reference.

Or is this more of a warning to users that they shouldn't use exti after nvic.enable?

Reply:
It' both. It's a warning, and also: Isn't it UB that two mutable references to the same data *exist*? (Regardless of them being used?)

Reply:
I'm not entirely sure. My guess is that it's not for the same reason that working with pointers is safe until you actually dereference them, which requires `unsafe` but I may be wrong

Reply:
With non lexical lifetime, it's safe if you don't use the mutable reference before enabling interrupt, as it would be accepted by the compiler to create another `&mut` after the enable line.

For examples, I prefer a static mut option to a static mut maybeuninit, less unsafe things to think about.